File: array/1409_queries_on_a_permutation_with_key/1409_queries_on_a_permutation_with_key.py
import collections

from typing import List



class Solution:
    def processQueries(self, queries: List[int], m: int) -> List[int]:
        tree = [0] * ((2*m) + 1)
        res = []
        
        def update(i,val):
            while i<len(tree):
                tree[i]+=val
                i+=(i&(-i))
    
        def prefixSum(i):
            s=0
            while i>0:
                s+=tree[i]
                i-=(i&(-i))
            return s
        
        hmap = collections.defaultdict(int)
        
        for i in range(1,m+1):
            hmap[i] = i+m
            update(i+m,1)

        for i in queries:
            res.append(prefixSum(hmap[i])-1)
            update(hmap[i],-1)
            update(m,1)
            hmap[i] = m
            m-=1

        return res


queries = [3,1,2,1]
m = 5
res = Solution().processQueries(queries, m)
print(res)

File: array/1409_queries_on_a_permutation_with_key/1409_queries_on_a_permutation_with_key.go
package main

import "fmt"

func processQueries(queries []int, m int) []int {
	var p = make([]int, m)
	var res = make([]int, len(queries))
	for i := 0; i < m; i++ {
		p[i] = i + 1
	}
	for index, value := range queries {
		//found value in p's index
		var i int
		for i = 0; i < len(p); i++ {
			if p[i] == value {
				res[index] = i
				break
			}
		}
		//move p
		tmp := p[i]
		copy(p[1:i+1], p[0:i])
		p[0] = tmp
	}
	return res
}

func main() {
	queries := []int{3, 1, 2, 1}
	m := 5
	res := processQueries(queries, m)
	fmt.Println(res)
}
